### PR TITLE
Support indicators in structured output and implement one for `build`…

### PR DESCRIPTION
**Pull request by mkaput**
… command

commit-id:d9ef14ef
Signed-off-by: Marek Kaput <marek.kaput@swmansion.com>

---

**Stack**:
- #1160
- #1166
- #1165
- #1154
- #1146 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: protostar/commands/build_command.py [hunk 1]
```diff
@@ -50,31 +65,26 @@ def arguments(self):
...
         ]
 
-    async def run(self, args: Namespace):
-        await self.build(
-            output_dir=args.compiled_contracts_dir,
-            disable_hint_validation=args.disable_hint_validation,
-            relative_cairo_path=args.cairo_path,
-        )
+    async def run(self, args: Any):
+        write = self._messenger_factory.from_args(args)
```
Comment (kasperski95) [inline protostar/commands/build_command.py:69]:
```suggestion
        messanger = self._messenger_factory.from_args(args)
```
It's strange `messanger_factory` produces `write`.

File: protostar/io/output.py [hunk 2]
```diff
@@ -39,7 +62,28 @@ def __call__(self, message: Message):
...
+
+        rendered = message_template.format_dict()
+
+        self._print({**rendered, "event": "started"})
+        try:
+            yield
+            self._print({**rendered, "event": "ok"})
+        except BaseException:
+            self._print({**rendered, "event": "failed"})
+            raise
```
Comment (kasperski95) [inline protostar/io/output.py:83]: (We need to figure out how to display errors (ProtostarException), but this is out of scope of this PR.)

**Review by kasperski95 — APPROVED** [review]
LGTM

File: protostar/commands/build_command.py (hunk 1 above)
Comment (mkaput) [reply protostar/commands/build_command.py:69]: Yeah, but it's also strange to have `messenger(...)` function (noun vs verb)

File: protostar/io/output.py (hunk 2 above)
Comment (mkaput) [reply protostar/io/output.py:83]: #1155

File: protostar/commands/build_command.py (hunk 1 above)
Comment (kasperski95) [reply protostar/commands/build_command.py:69]: `write_fn_factory`? I don't know. Let's leave it as it is.

File: protostar/io/output.py [hunk 3]
```diff
@@ -39,7 +62,28 @@ def __call__(self, message: Message):
                 f"JSON output is not supported for messages of type {type(message).__name__}."
             )
 
-        dictionary = message.format_dict()
-        json.dump(dictionary, sys.stdout, allow_nan=False, separators=(",", ":"))
+        obj = message.format_dict()
+        self._print(obj)
+
+    @contextmanager
+    def activity(self, message_template: Message) -> Generator[None, None, None]:
```
Comment (Arcticae) [inline protostar/io/output.py:69]: I'm wondering if that's something that users would like to see in the output... I think the activity indicator thing is useful for having some UX thing making people aware that something is happening, whilst when having the json output i'd rather have the output in one plain json.

Comment (mkaput) [reply protostar/io/output.py:69]: I think I have a good reason for having these messages: measuring timing of the process in the tooling. With these messages, one is able to determine time cost of compilation minus Protostar booting time.

But nonetheless: I just copied what `cargo test` does in its JSON output. Maybe it doesn't really make sense here.

Another point of this PR was to move activity indicators to structured output framework. Maybe this command could simply skip JSON output as for now, like we did for `protostar format`?

Comment (Arcticae) [reply protostar/io/output.py:69]: I think we can skip, since the output of the build is not that important - we know the preliminary contract names (and thus output names)

Comment (mkaput) [reply protostar/io/output.py:69]: done

**Review by Arcticae — APPROVED** [review]
